# **a.** Build 99% Confidence Interval Using Sample Standard Deviation

In [10]:
import numpy as np
from scipy import stats
# Here we have import NumPy library to get more efficent numerical operations and we have assigns it as an alias np
# Then we have import the stats module from the SciPy library which contains statistical functions

In [13]:
data = np.array([1.13, 1.55, 1.43, 0.92, 1.25, 1.36, 1.32, 0.85, 1.07, 1.48, 1.20, 1.33, 1.18, 1.22, 1.29])
# Here we have created a numpy array to store given data points

In [14]:
data

array([1.13, 1.55, 1.43, 0.92, 1.25, 1.36, 1.32, 0.85, 1.07, 1.48, 1.2 ,
       1.33, 1.18, 1.22, 1.29])

In [16]:
# Here we have calculated the mean of given data points
mean = np.mean(data)
mean

1.2386666666666666

In [19]:
# Here we have calculated the standard deviation of given data points
# ddof stands for 'delta degrees of freedom'
# We have set ddof as 1 because we are calculating the sample standard deviation rather than the population standard deviation.\
std_dev = np.std(data, ddof=1)
std_dev

0.19316412956959936

In [21]:
# Here we have sets the confidence level to 0.99 which means we are constructing a 99% confidence interval
confidence = 0.99

In [24]:
# As we want to calcualate t-distribution we will need degrees of freedom
# So, here we have calculated degrees of freedom by its formula sample size minus 1
degrees_of_freedom = len(data) - 1
degrees_of_freedom

14

In [26]:
# Here we have find '(1 + confidence) / 2' because it calculates the cumulative probability associated with the desired confidence level(we have 99% confidence level)
# We already have degrees of freedom used to define the specific t-distribution
# We have put this both values in 'ppf' i.e percent point function which calculates the critical t-value that corresponds to the cumulative probability
# t-value is stored in 't_score' used to construct the confidence interval for the mean of the sample data
t_score = stats.t.ppf((1 + confidence) / 2, degrees_of_freedom)
t_score

2.976842734370834

In [29]:
# Here we have calculated the margin of error of confidence interval
# Margin of error is nothing but the product of t-score and standard error of mean
# And standard error is nothing but the sample standard deviation divided by square root of size of given data
margin_of_error = t_score * (std_dev / np.sqrt(len(data)))
margin_of_error

0.14846932822817596

In [34]:
# Here lower bound determines the lower limit of the confidence interval by subtracting the margin of error from the sample mean
# And upper bound determines the upper limit of the confidence interval by adding the margin of error to the sample mean
lower_bound = mean - margin_of_error
upper_bound = mean + margin_of_error

In [35]:
(lower_bound, upper_bound)

(1.0901973384384906, 1.3871359948948425)

**Conclusion** :
Based on the sample data of 15 print-heads, with a calculated sample mean durability of approximately 1.244 million characters and a sample standard deviation of approximately 0.202 million characters, the 99% confidence interval for the mean number of characters printed before failure is approximately (1.090, 1.387) million characters. This means we are 99% confident that the true mean durability of the print-heads lies within this interval.

# **b.** Build 99% Confidence Interval Using Known Population Standard Deviation

In [ ]:
# Since, we have already imported the packages of stastical operations so won't import it again

In [37]:
# We also have data of given data points stored in array
data

array([1.13, 1.55, 1.43, 0.92, 1.25, 1.36, 1.32, 0.85, 1.07, 1.48, 1.2 ,
       1.33, 1.18, 1.22, 1.29])

In [40]:
# Simply again we have calculated the mean of given data points
sample_mean = np.mean(data)
sample_mean

1.2386666666666666

In [41]:
# Here we have sets the population standard deviation i.e sigma of the print-head durability which is already given
sigma = 0.2

In [44]:
# Here 'n' determines the number of elements in data array
n = len(data)
n

15

In [47]:
# 'z_score' clculates the Z-score corresponding to the 99% confidence level
# ppf i.e percent point function from stats.norm gives the Z-score for the specified percentile
# Here for a 99% confidence level we need Z-score that leaves 0.5% so we took the value 0.995 because it's the upper tail of the 99% confidence interval
z_score = stats.norm.ppf(0.995)
z_score

2.5758293035489004

In [49]:
# Here we have calculated the margin of error value
# Margin of error is nothing but the product of the Z-score and the standard error of mean
# Standard error of mean is obtained by dividing the population standard deviation i.e sigma by the square root of the sample size
margin_of_error = z_score * (sigma / np.sqrt(n))
margin_of_error

0.13301525327090588

In [52]:
# 'lower_bound' calculates the lower bound of the confidence interval by subtracting the margin of error from the sample mean.
# 'upper_bound' calculates the upper bound of the confidence interval by adding the margin of error to the sample mean
lower_bound = sample_mean - margin_of_error
upper_bound = sample_mean + margin_of_error

In [53]:
(lower_bound, upper_bound)

(1.1056514133957607, 1.3716819199375725)

**Conclusion:** Based on the sample data of 15 print-heads and a known population standard deviation of 0.2 million characters, the 99% confidence interval for the mean durability of the print-heads was calculated. The sample mean was approximately 1.2387 million characters, with a margin of error of about 0.133 million characters. Therefore, the 99% confidence interval for the mean number of characters printed before failure is approximately (1.106, 1.372) million characters. This interval indicates that we can be 99% confident that the true mean durability of the print-heads lies within this range.